### PR TITLE
fix: Fix miner

### PR DESCRIPTION
**Pull request by driftluo**
### What problem does this PR solve?

1. In the case that the poll interval is small enough, the miner may work on unknown work, which is equivalent to doing useless work(work cache is 32, the channel is unbound, worker get new work just one loop one get)
2. Assuming that miner must be submitted successfully 
3. Miner will submit uncle block 
4. If `block_on_submit` is false, the miner will not submit the block

### Tests

- setup a dev node(test for Dummy)，use miner subcommand for it
- setup a staging node(test for Eaglesong)，use miner subcommand for it
- change `block_on_submit` to false，and `poll_interval` = 200, old miner can't work, the new one works well
- old miner may have uncle block submit, the new one doesn't

### Related changes

Rewritten part of the miner logic 

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by doitian** [comment]
This PR requires the tests check list section in the template:

```
Tests <!-- At least one of them must be included. -->

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code (skip ci)
```

**Review by doitian — CHANGES_REQUESTED** [review]
Tests needed

**Comment by driftluo** [comment]
> Tests needed

Added manual test steps

File: miner/src/miner.rs [hunk 1]
```diff
@@ -82,26 +82,26 @@ impl Miner {
...
                     Ok(work) => {
-                        let pow_hash= work.block.header().calc_pow_hash();
-                        let (target, _,) = compact_to_target(work.block.header().raw().compact_target().unpack());
-                        self.works.put(pow_hash.clone(), work);
-                        self.notify_workers(WorkerMessage::NewWork{pow_hash, target});
+                        match work {
+                            Works::FailSubmit(hash) => {
+                                self.legacy_work.pop(&hash);
+                            },
+                            Works::New(work) => self.submit_work(work),
```
Comment (quake) [inline miner/src/miner.rs:94]: This fn name is very misleading, suggest to rename to `notify_new_work`
```suggestion
                            Works::New(work) => self.notify_new_work(work),
```

File: miner/src/miner.rs [hunk 2]
```diff
@@ -115,43 +115,81 @@ impl Miner {
...
-                .build();
-            let block_hash = block.hash();
-            if self.stderr_is_tty {
-                debug!("Found! #{} {:#x}", block.number(), block_hash);
-            } else {
-                info!("Found! #{} {:#x}", block.number(), block_hash);
-            }
+    fn submit_work(&mut self, work: Work) {
+        let parent_hash = work.block.header().into_view().parent_hash();
+        if !self.legacy_work.contains(&parent_hash) {
```
Comment (quake) [inline miner/src/miner.rs:120]: I don't think the miner will work on the updated block_template after this change, could you confirm that?

```
get_block_template => got template A
miner working on template A
get_block_template => got template B (with same parent hash as A)
miner still working on template A
```

Comment (driftluo) [reply miner/src/miner.rs:120]: legacy work only insert on submitting block successfully, the template corresponding to the block that has not been submitted will be updated in real-time

Comment (quake) [reply miner/src/miner.rs:120]: Thanks for the answer, but I think there's still an issue here: when the miner submits work with an error, the parent_hash 
 will be added to legacy_work: https://github.com/nervosnetwork/ckb/blob/3389d741c8246aea90936c9b00031b972ee9393f/miner/src/client.rs#L189 then the miner can't continue to work properly, I think we need to distinguish whether the error type is a duplicate commit or other error. I think you can simulate a network error during miner submitting block by restarting the ckb node, and you will see that miner stops working after the ckb node restart.

Comment (quake) [reply miner/src/miner.rs:120]: I think it's a miner issue and will approve this PR.

**Review by quake — DISMISSED** [review]

File: miner/src/miner.rs (hunk 2 above)
Comment (driftluo) [reply miner/src/miner.rs:120]: When submitting asynchronously, if an error occurs, a submit fail error is returned, and the parent that is considered to have been successfully submitted asynchronously will be removed
```
Works::FailSubmit(hash) => {
    self.legacy_work.pop(&hash);
},
```

Comment (quake) [reply miner/src/miner.rs:120]: This are the steps to re-produce:

```
ckb init -c dev --ba-arg 0xed99d08d751ae32df7fc6f3048b7cca56c3bfafa
ckb run
ckb miner
```

ctrl+c to stop the ckb process and restart it
```
^C
ckb run
```

ckb miner stops:

```
2021-08-17 11:41:33.041 +08:00 main INFO ckb_miner::miner  block 66 pow_hash: 0xbe4eefda696a8b579518446e4957dd44ce7918856d0b9c643e079133389c6b33, header: HeaderView { data: Header { raw: RawHeader { version: Uint32(0x00000000), compact_target: Uint32(0x00000120), timestamp: Uint64(0x243533527b010000), number: Uint64(0x4200000000000000), epoch: Uint64(0x0000004200e80300), parent_hash: Byte32(0x68719de4aad252c2ddc10787f9e55647569f7168de5fa645f37eb7d84c34b073), transactions_root: Byte32(0xe5817d444e309fb6dd1b83610b87bdd03636f70efd782f9c607308626e67c284), proposals_hash: Byte32(0x00000000000000Found! #66 0x03447d0917ffee80ea1bc6765aff52984f0a1c64ecfc34f9e0aaf5eb72f27b12
Dummy-Worker ⠁ [00:00:00] 
Total nonces found:  14
2021-08-17 11:41:33.246 +08:00 main INFO ckb_miner::miner  block 67 pow_hash: 0xdb83705bec7171a27f86f57fd070a3ae060f00cb2f9888ac905ebe97ab093f37, header: HeaderView { data: Header { raw: RawHeader { version: Uint32(0x00000000), compact_target: Uint32(0x00000120), timestamp: Uint64(0xf33533527b010000), number: Uint64(0x4300000000000000), epoch: Uint64(0x0000004300e80300), parent_hash: Byte32(0x03447d0917ffee80ea1bc6765aff52984f0a1c64ecfc34f9e0aaf5eb72f27b12), transactions_root: Byte32(0x70c52a4b0c419dd6ca31fd53f8a46edf4f0a7416a0ff320e08cbec264eefce59), proposals_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), extra_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), dao: Byte32(0x1063a854d92da12ebb312949f5862300ea7296003c03000000f1e7d450fbfe06) }, nonce: Uint128(0x63801d78dcd838d38b482bbd64b6d60e) }, hash: Byte32(0x8cf7a704c8ec21cc02304e912c61b9318d66a5a7aa0513efb75528ea38d6dc3a) }
2021-08-17 11:41:33.247 +08:00 main ERROR ckb_miner::miner  rpc call submit_block error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 61, kind: ConnectionRefused, message: "Connection refused" })))
Found! #67 0x8cf7a704c8ec21cc02304e912c61b9318d66a5a7aa0513efb75528ea38d6dc3a
Dummy-Worker ⠁ [00:00:00] 
Total nonces found:  15
```

Could you confirm that it's working properly on your machine?

Comment (driftluo) [reply miner/src/miner.rs:120]: This kind of scene works normally on my machine, no matter it is the debug version or the release version, it cannot be reproduced. 
```
2021-08-17 12:08:20.156 +08:00 main INFO ckb_miner::miner  block 139 pow_hash: 0xb2814835918fd056fd3f2ad1f55ae6526b08b6f742df000d00aa1e4071808a0f, header: HeaderView { data: Header { raw: RawHeader { version: Uint32(0x00000000), compact_target: Uint32(0x00000120), timestamp: Uint64(0xc6b94b527b010000), number: Uint64(0x8b00000000000000), epoch: Uint64(0x0000008b00e80300), parent_hash: Byte32(0xd7b2477487e7c64674c52c1e04a797b37ebfba835fb2b78e22c4b894937c4913), transactions_root: Byte32(0x3c9fbd503836f6bb986d81d99d42ae0cb1be0e0184b993913f8808f9c8c839da), proposals_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), extra_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), dao: Byte32(0x305c89196d3ea12e4988ff58f8862300c400fe70a606000000194417b7fbfe06) }, nonce: Uint128(0xea20f3ceec727098be9dbae2d3aec05e) }, hash: Byte32(0xfd90ed98afd8047babb9ce0b94f73b642283acd90d3bc569583ef73b127adca6) }
2021-08-17 12:08:20.158 +08:00 GlobalRt-6 ERROR ckb_miner::client  rpc call submit_block error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:20.158 +08:00 main ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:20.261 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:20.462 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:20.663 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:20.864 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:21.065 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:21.266 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:21.467 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:21.667 +08:00 client ERROR ckb_miner::client  rpc call get_block_template error: Http(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })))
2021-08-17 12:08:23.578 +08:00 main INFO ckb_miner::miner  block 139 pow_hash: 0x1923b2950b740ad664731125b8ef0fe598cc3dd2b86adcf9ebbe64f263dd483e, header: HeaderView { data: Header { raw: RawHeader { version: Uint32(0x00000000), compact_target: Uint32(0x00000120), timestamp: Uint64(0x23c74b527b010000), number: Uint64(0x8b00000000000000), epoch: Uint64(0x0000008b00e80300), parent_hash: Byte32(0xd7b2477487e7c64674c52c1e04a797b37ebfba835fb2b78e22c4b894937c4913), transactions_root: Byte32(0x3c9fbd503836f6bb986d81d99d42ae0cb1be0e0184b993913f8808f9c8c839da), proposals_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), extra_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), dao: Byte32(0x305c89196d3ea12e4988ff58f8862300c400fe70a606000000194417b7fbfe06) }, nonce: Uint128(0xdd3d9ad42b368fb4a78269a23be8c36b) }, hash: Byte32(0xf6b339e696a80e0c0c218d502f4f1d165098b3f65eeb8c2ccd45e94ab73ab475) }
```

Comment (driftluo) [reply miner/src/miner.rs:120]: solve by https://github.com/nervosnetwork/ckb/pull/2947/commits/78bec450927bcef02c44a55c590046a0e176c71f，set poll interval always 1000ms

**Review by quake — APPROVED** [review]

**Review by yangby-cryptape — APPROVED** [review]

**Comment by yangby-cryptape** [comment]
bors merge=quake,yangby-cryptape

**Pull request merged**